cupyx/scipy/spatial/delaunay_2d/_tri.py

```diff
@@ -0,0 +1,570 @@
+        self.values = cupy.empty(self.n_points, dtype=cupy.int32)
+        self.vert_tri = cupy.zeros(self.n_points, dtype=cupy.int32)
+
+        self._org_flip_num = []
+
+    @property
+    def counters(self):
+        return self._counters[self.counters_offset:]
+
+    def _renew_counters(self):
+        self.counters_offset += self.counters_size
+        if self.counters_offset > 8192:
+            self.counters_offset = 0
+            self._counters[:] = 0
+
+    def _construct_initial_triangles(self):
+        # Find extreme points in the x-axis
+        v0 = cupy.argmin(self.point_vec[:-1, 0])
+        v1 = cupy.argmax(self.point_vec[:-1, 0])
+
+        # Find furthest point from v0 and v1, a.k.a the biggest
+        # triangle available
+        compute_distance_2d(self.point_vec, v0, v1, self.values)
+
+        v2 = cupy.argmax(self.values[:-1])
+
+        # Check if the three points are not coplanar
+        ori = _check_if_coplanar_points(self.point_vec, v0, v1, v2)
+
+        is_coplanar = cupy.where(ori == 0.0, True, False)
+        if is_coplanar:
+            raise ValueError(
+                'The input is degenerate, the extreme points are close to '
+                'coplanar')
+
+        tri_ort = _compute_triangle_orientation(ori)
+        tri = cupy.r_[v0, v1].astype(cupy.int32)
+        tri = cupy.where(tri_ort == -1, tri[::-1], tri)
+
+        # Create the initial triangulation
+        # Compute the centroid of v0 v1 v2, to be used as the kernel point.
+        tri = cupy.r_[tri, v2].astype(cupy.int32)
+        self.point_vec[-1] = self.point_vec[tri].mean(0)
+
+        self.pred_consts = cupy.empty(18, cupy.float64)
+        init_predicate(self.pred_consts)
+
+        # Put the initial triangles at the Inf list
+        make_first_tri(
+            self.triangles, self.triangle_opp, self.triangle_info,
+            tri, self.n_points - 1)
+
+        self._renew_counters()
+
+        exact_check = cupy.empty(self.n_points, dtype=cupy.int32)
+        init_point_location_fast(
+            self.vert_tri, self.n_points, exact_check,
+            self.counters, tri, self.n_points - 1, self.point_vec,
+            self.points_idx, self.pred_consts)
+
+        init_point_location_exact(
+            self.vert_tri, self.n_points, exact_check,
+            self.counters, tri,
+            self.n_points - 1, self.point_vec, self.points_idx,
+            self.pred_consts)
+
+        self.available_points = self.n_points - 4
+        self.tri_num = 4
+
+    def _init_for_flip(self):
+        min_val = self.points.min()
+        max_val = self.points.max()
+        range_val = max_val - min_val
+
+        # Sort the points spatially according to their Morton numbers
+        get_morton_number(self.points, self.n_points - 1, min_val,
+                          range_val, self.values)
+
+        self.values[-1] = 2 ** 31 - 1
+        unique_values, unique_index = cupy.unique(
+            self.values, return_index=True)
```

Comment:
This unique is also a bottleneck. This is an issue with the original implementation regarding duplicate points, at some point in the reference C++ implementation looped infinitely when trying to insert duplicate points. I haven't tried to remove it